### DISCUSSION TITLE
New JavaScript API methods

### DISCUSSION THREAD

**Discussion by alexiglesias93** [finsweet/wized-roadmap#8] [RFCs]
There are a few API methods that have been requested by the community.
Let's discuss them!

## Exposing App SDK's via `Wized.apps`
**Update:**
We have created a better approach via Wized.requests.getClient().
Here's the preview documentation for the new method, feel free to give it a read an provide feedback!
https://wized-docs-2-emcelod7u-finsweet.vercel.app/javascript-api/#wized-requests-getclient

--- 

**Original:**
This one is by far the most requested one.
In Wized there are some integrations like [Supabase](https://supabase.com/docs/reference/javascript/start) and [Firebase](https://firebase.google.com/docs/reference/js) that have their own native SDK. Although Wized tries to stick as close as possible to their API, there are some missing features that are not built into Wized (yet).
To have full control of the Apps, users have asked if it would be possible to access those SDK's via the Wized JavaScript API.
I'm thinking we could do something like this:
```js
Wized.apps.appName.appMethod();
```
For example, the user has a Supabase app called `my_supabase`, accessing the Supabase SDK [to perform a basic select](https://supabase.com/docs/reference/javascript/select) would look like this:
```js
const { data, error } = await Wized.apps.my_supabase.from('countries').select();
```
### Pros
- Will give developers 100% control of the SDK clients, allowing for full extensibility.
### Cons
- Wized currently only loads parts of the SDKs that are actively used, and does [tree shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking) when possible. Exposing the SDKs through the public API would mean having to include many methods that may never be used by the users, thus making the downloaded resources bigger than necessary.
- In addition, Wized only loads the SDKs when a request that uses them is performed, this way the amount of resources downloaded in each page are the minimum necessary. By exposing the SDKs via our public JS API we would have to rethink this strategy as they would have to be always available.

An alternative that could potentially fix the latter problem could be to expose a single method to retrieve an SDK, like:
```js
Wized.apps.getClient('app_name');
```
So the previous example would look like:
```js
const supabaseClient = await Wized.apps.getClient('my_supabase');
const { data, error } = await supabaseClient.from('countries').select();
```

I'm open to ideas!

## Reactive functions via `Wized.data.effect()`
Wized uses [@vue/reactivity](https://www.npmjs.com/package/@vue/reactivity) under the hood to implement granular reactivity using signals.
The base of Wized's signals is the [effect()](https://github.com/vuejs/core/blob/main/packages/reactivity/src/effect.ts#L193) method, which allows us to create reactive effects that re-run any time its dependencies change.
Exposing this method in our API would allow users to implement more advanced custom logic that smartly reacts to their Wized application state.
Here's an example:
```js
Wized.data.effect(() => {
  // Will re-run any time v.count changes
  console.log(v.count);
});
```

## Aborting requests via `Wized.requests.abort()`
Sometimes users want to abort an ongoing request. This would be possible only when using REST requests (via an [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)) and Supabase requests (via [abortSignal](https://supabase.com/docs/reference/javascript/db-abortsignal)), but it's worth exposing it in my opninion.
It would look something like this:
```js
Wized.requests.abort('my_request');
```
I don't know if it would also be valuable to include this request abortion control in the Wized native event after actions. Same as users can `Perform request`, we could add a new `Abort request` action. Thoughts?

## Wized.destroy() or Wized.restart() methods for SPA mode
Some users tried to use Wized in combination with an SPA router like [barba.js](https://barba.js.org/) or [swup](https://swup.js.org/). These methods would allows users to safely destroy and restart Wized's runtime to make it compatible with SPA navigations.

**Comment by dima-dinero** [comment]
Wized.apps would be a gamechanger, especially working with Supabase. There's a lot of requests like: uploading multiple files to a bucket, getting the list of bucket's items, or delete file etc. that can't be done natively in Wized. And I'm really suffering right now, working on my project 😄 So, I'll be very gratefull to you for this feature)

**Comment by alexiglesias93** [comment]
Update on `Wized.apps`: we have created a better approach via `Wized.requests.getClient()`.
Here's the preview documentation for the new method, feel free to give it a read an provide feedback!
https://wized-docs-2-emcelod7u-finsweet.vercel.app/javascript-api/#wized-requests-getclient
